File: training/objectivec/CellarsAndCentaurs/Domain/Model/ArmorClass.m
#import "ArmorClass.h"
#import "CellarsAndCentaurs.h"
#include "CreatureLimits.h"

@implementation ArmorClass

- (instancetype)initWithBase:(NSInteger)base bonus:(NSInteger)bonus {
    return [self initWithBase:base bonus:bonus description:CreatureUtil.standardArmorClassDescription];
}

- (instancetype)initWithBase:(NSInteger)base bonus:(NSInteger)bonus description:(NSString *)armorDescription {
    self = [super init];
    if (self) {
        _base = base;
        _bonus = bonus;
        _armorDescription = [armorDescription copy];
    }
    return self;
}

- (NSInteger)total {
    return self.base + self.bonus;
}

- (BOOL)isWithinLimits {
    return self.total >= MIN_ARMOR_CLASS && self.total <= MAX_ARMOR_CLASS;
}

@end

File: sample-projects/swift/Sources/CellarsAndCentaurs/Domain/Model/ArmorClass.swift
struct ArmorClass {
    let description: String
    var base: Int
    var bonus: Int

    var total: Int { base + bonus }

    init(base: Int, bonus: Int, description: String = CreatureUtil.STANDARD_ARMOR_CLASS_DESCRIPTION) {
        self.base = base
        self.bonus = bonus
        self.description = description
    }
}
